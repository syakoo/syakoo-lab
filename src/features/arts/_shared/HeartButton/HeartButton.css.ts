import { ComplexStyleRule, keyframes, style } from "@vanilla-extract/css";
import { recipe } from "@vanilla-extract/recipes";
import { tokens } from "@/design-system/tokens";
import { range } from "@/utils/array/range";

const circleOption = {
  /**
   * アニメーション用の円の数
   */
  circleNum: 6,
};

const fadeIn = keyframes({
  "0%": {
    opacity: 0,
    transform: "scale(0) rotate(0deg)",
  },
  "30%": {
    opacity: 0,
  },
  "60%": {
    opacity: 1,
  },
  "100%": {
    opacity: 1,
    transform: "scale(1.3) rotate(180deg)",
  },
});

const fadeOut = keyframes({
  "0%": {
    opacity: 1,
  },
  "100%": {
    opacity: 0,
  },
});

const bounce = keyframes({
  "0%": {
    transform: "scale(1)",
  },
  "33%": {
    transform: "scale(0.3)",
  },
  "66%": {
    transform: "scale(1.5)",
  },
  "90%": {
    transform: "scale(0.8)",
  },
  "95%": {
    transform: "scale(1.1)",
  },
  "100%": {
    transform: "scale(1)",
  },
});

export const heartButtonStyles = {
  root: style({
    position: "relative",
    display: "flex",
    width: "100%",
    padding: "5px 4px 3px 4px",
    color: tokens.colors.text.secondary,
    background: "transparent",
    border: "none",
    borderRadius: tokens.radii[1000],

    ":disabled": {
      color: tokens.colors.accent.error.foreground,
    },
    selectors: {
      "&:not(:disabled):hover": {
        cursor: "pointer",
        background: tokens.colors.text.tertiary,
        transition: "0.4s",
      },
      "&:not(:disabled):active": {
        filter: "brightness(0.7)",
      },
    },
  }),
  iconWrapper: recipe({
    base: {
      display: "flex",
      width: "100%",
      aspectRatio: "1 / 1",
    },

    variants: {
      animationFired: {
        true: {
          animation: `${bounce} 0.8s ease-in-out`,
        },
      },
    },
  }),
  animationCircleContainer: style({
    position: "absolute",
    top: "0",
    left: "0",
    width: "100%",
    height: "100%",
    transformOrigin: "center",
    opacity: 0,
    animation: `${fadeIn} 0.6s 0.4s ease-in-out forwards, ${fadeOut} 0.2s 0.8s ease-in-out forwards`,
  }),
  animationCircle: recipe({
    base: {
      position: "absolute",

      minWidth: 4,
      width: "15%",
      aspectRatio: "1 / 1",
      borderRadius: tokens.radii[1000],

      selectors: {
        "&:nth-child(4n)": {
          backgroundColor: tokens.colors.accent.error.foreground,
        },
        "&:nth-child(4n+1)": {
          backgroundColor: tokens.colors.accent.success.foreground,
        },
        "&:nth-child(4n+2)": {
          backgroundColor: tokens.colors.accent.info.foreground,
        },
        "&:nth-child(4n+3)": {
          backgroundColor: tokens.colors.accent.warn.foreground,
        },
      },
    },

    variants: {
      index: range(0, circleOption.circleNum).reduce<
        Record<number, ComplexStyleRule>
      >(
        (p, index) => ({
          ...p,
          [index]: {
            top: `calc(50% * ${Math.sin(
              (index / circleOption.circleNum) * Math.PI * 2,
            )} + 50%)`,
            left: `calc(50% * ${Math.cos(
              (index / circleOption.circleNum) * Math.PI * 2,
            )} + 50%)`,
            // NOTE: 小さい円の半径分戻す
            transform: "translate(-50%, -50%)",
          },
        }),
        {},
      ),
    },
  }),
};
