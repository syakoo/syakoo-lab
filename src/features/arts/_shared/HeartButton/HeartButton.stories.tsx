import type { Meta, StoryObj } from "@storybook/react";
import { userEvent, within } from "@storybook/testing-library";

import { HeartButton } from ".";

const meta = {
  component: HeartButton,
  decorators: [
    (Story) => (
      <div style={{ width: "36px" }}>
        <Story />
      </div>
    ),
  ],
  parameters: { testLevel: "snapshot" },
} satisfies Meta<typeof HeartButton>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {};

export const Activated: Story = {
  args: {
    initialValue: true,
  },
};

export const ClickButton: Story = {
  args: {
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    onClick: () => {},
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    const button = canvas.getByRole("button");

    await userEvent.click(button);
  },
};
