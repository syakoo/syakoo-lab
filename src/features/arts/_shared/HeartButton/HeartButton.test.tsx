import { testStories } from "@/test/testStories";

import * as stories from "./HeartButton.stories";

describe("HeartButton", () => {
  testStories(stories);
});
