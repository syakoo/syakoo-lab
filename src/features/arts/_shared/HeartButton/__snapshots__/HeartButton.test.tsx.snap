// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeartButton renders Activated 1`] = `
<body>
  <div>
    <div
      style="width: 36px;"
    >
      <button
        class="HeartButton_heartButtonStyles_root__ap10oo3"
        disabled=""
        type="button"
      >
        <div
          class="HeartButton_heartButtonStyles_iconWrapper__ap10oo4"
        >
          <svg
            fill="currentColor"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M21 8.25c0-2.485-2.099-4.5-4.688-4.5-1.935 0-3.597 1.126-4.312 2.733-.715-1.607-2.377-2.733-4.313-2.733C5.1 3.75 3 5.765 3 8.25c0 7.22 9 12 9 12s9-4.78 9-12z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
      </button>
    </div>
  </div>
</body>
`;

exports[`HeartButton renders ClickButton 1`] = `
<body>
  <div>
    <div
      style="width: 36px;"
    >
      <button
        class="HeartButton_heartButtonStyles_root__ap10oo3"
        disabled=""
        type="button"
      >
        <div
          class="HeartButton_heartButtonStyles_animationCircleContainer__ap10oo6"
        >
          <div
            class="HeartButton_heartButtonStyles_animationCircle__ap10oo7 HeartButton_heartButtonStyles_animationCircle_index_0__ap10oo8"
          />
          <div
            class="HeartButton_heartButtonStyles_animationCircle__ap10oo7 HeartButton_heartButtonStyles_animationCircle_index_1__ap10oo9"
          />
          <div
            class="HeartButton_heartButtonStyles_animationCircle__ap10oo7 HeartButton_heartButtonStyles_animationCircle_index_2__ap10ooa"
          />
          <div
            class="HeartButton_heartButtonStyles_animationCircle__ap10oo7 HeartButton_heartButtonStyles_animationCircle_index_3__ap10oob"
          />
          <div
            class="HeartButton_heartButtonStyles_animationCircle__ap10oo7 HeartButton_heartButtonStyles_animationCircle_index_4__ap10ooc"
          />
          <div
            class="HeartButton_heartButtonStyles_animationCircle__ap10oo7 HeartButton_heartButtonStyles_animationCircle_index_5__ap10ood"
          />
        </div>
        <div
          class="HeartButton_heartButtonStyles_iconWrapper__ap10oo4 HeartButton_heartButtonStyles_iconWrapper_animationFired_true__ap10oo5"
        >
          <svg
            fill="currentColor"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M21 8.25c0-2.485-2.099-4.5-4.688-4.5-1.935 0-3.597 1.126-4.312 2.733-.715-1.607-2.377-2.733-4.313-2.733C5.1 3.75 3 5.765 3 8.25c0 7.22 9 12 9 12s9-4.78 9-12z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
      </button>
    </div>
  </div>
</body>
`;

exports[`HeartButton renders Default 1`] = `
<body>
  <div>
    <div
      style="width: 36px;"
    >
      <button
        class="HeartButton_heartButtonStyles_root__ap10oo3"
        type="button"
      >
        <div
          class="HeartButton_heartButtonStyles_iconWrapper__ap10oo4"
        >
          <svg
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M21 8.25c0-2.485-2.099-4.5-4.688-4.5-1.935 0-3.597 1.126-4.312 2.733-.715-1.607-2.377-2.733-4.313-2.733C5.1 3.75 3 5.765 3 8.25c0 7.22 9 12 9 12s9-4.78 9-12z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
      </button>
    </div>
  </div>
</body>
`;
